src/avrpart.c

```diff
@@ -529,68 +496,165 @@ AVRMEM_ALIAS *avr_find_memalias(const AVRPART *p, const AVRMEM *m_orig) {
+    int m_size[] = {0, m->size, m->page_size, m->offset};
+    const int m_base[] = {0, 10, 10, 16};
+
+    // Mem desc charcter length
+    AVRMEM_ALIAS *a = avr_find_memalias(p, m);
+    const char *m_desc_a = a? a->desc: "";
+    int cnt = strlen(m->desc) + strlen(a? "/": "") + strlen(m_desc_a);
+    if(m_char_max[0] < cnt)
+      m_char_max[0] = cnt;
+    if(m_char_max[0] < (int)strlen(table_colum[0]))
+      m_char_max[0] = strlen(table_colum[0]);
+    // Mem size/pgsize/offset character length
+    for(int i = 1; i < 4; i++) {
+      cnt = 0;
+      do {
+        m_size[i] /= m_base[i];
+        ++cnt;
+      } while (m_size[i] != 0);
+      if(m_char_max[i] < cnt)
+        m_char_max[i] = cnt;
+      if(m_char_max[i] < (int)strlen(table_colum[i]))
+        m_char_max[i] = strlen(table_colum[i]);
+    }
+  }
+  m_char_max[3] += strlen("0x");
+
+  // Print memory table header
+  if(p->prog_modes & (PM_PDI | PM_UPDI)) {
+    fprintf(f,
+      "\n%s%-*s  %*s  %-*s  %*s\n"
+      "%s%*.*s--%*.*s--%*.*s--%*.*s\n",
+      prefix,
+      m_char_max[0], table_colum[0],
+      m_char_max[1], table_colum[1],
+      m_char_max[2], table_colum[2],
+      m_char_max[3], table_colum[3],
+      prefix,
+      m_char_max[0], m_char_max[0], table_padding,
+      m_char_max[1], m_char_max[1], table_padding,
+      m_char_max[2], m_char_max[2], table_padding,
+      m_char_max[3], m_char_max[3], table_padding);
+  } else {
+    fprintf(f,
+      "\n%s%-*s  %*s  %-*s\n"
+      "%s%*.*s--%*.*s--%*.*s\n",
+      prefix,
+      m_char_max[0], table_colum[0],
+      m_char_max[1], table_colum[1],
+      m_char_max[2], table_colum[2],
+      prefix,
+      m_char_max[0], m_char_max[0], table_padding,
+      m_char_max[1], m_char_max[1], table_padding,
+      m_char_max[2], m_char_max[2], table_padding);
   }
 
-  if (m != NULL) {
+  for (LNODEID ln=lfirst(p->mem); ln; ln=lnext(ln)) {
+    AVRMEM *m = ldata(ln);
     // Only print memory section if the previous section printed isn't identical
     if(prev_mem_offset != m->offset || prev_mem_size != m->size || str_eq(p->family_id, "")) {
       prev_mem_offset = m->offset;
       prev_mem_size = m->size;
-      AVRMEM_ALIAS *ap = avr_find_memalias(p, m);
-      /* Show alias if the current and the next memory section has the same offset
-      and size, we're not out of band and a family_id is present */
-      const char *mem_desc_alias = ap? ap->desc: "";
-      fprintf(f,
-              "%s%-11s %-8s %4d %5d %5d %4d %-6s %6d %4d %6d %5d %5d 0x%02x 0x%02x\n",
-              prefix,
-              m->desc,
-              mem_desc_alias,
-              m->mode, m->delay, m->blocksize, m->pollindex,
-              m->paged ? "yes" : "no",
-              m->size,
-              m->page_size,
-              m->num_pages,
-              m->min_write_delay,
-              m->max_write_delay,
-              m->readback[0],
-              m->readback[1]);
+
+      // Workaround to get the 0x prefix where it should be
```

Comment:
This code works but it is hard to read and/or maintain.

For example, it doesn't look good to print `0x0` as offset. Numbers between 0 and 9 should be printed 0 ... 9, not 0x0 ... 0x9. Making changes with that effect are rather hard in the current code. It is rare, but not impossible that some of the to be printed numbers are -1 (meaning unknown); they would currently not look good. I would suggest having two format specifiers for each number, one for numbers smaller than 10 (including negative ones) and another one for 10 or greater. Here are helper functions that return malloc'd space and their length:

```
static char *print_num(const char *fmt, int n) {
  return str_sprintf(n<10? "%d": fmt, n);
}
  
static int num_len(const char *fmt, int n) {
  char *p = print_num(fmt, n);
  int ret = strlen(p);
  free(p);

  return ret;
}
```

Usage would be sth like

```
  int len = num_len("0x%04x", mem->offset);
  if(len > max_len)
    max_len = len;
```

for computing the max width of a column; and for printing it would be sth like

```
   char *moff = print_num("0x%04x", mem->offset);
   printf(" ... %*s ...", max_len, moff);
   free(moff);
```

Reply:
Now _that's_ clever! Let me see what I can do...